### PR TITLE
43 distinct subsequences

### PR DESCRIPTION
**Pull request by Mohamed-Elnageeb**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Zeinab15 — COMMENTED** [review]
The code implementation is efficient, well-structured, and robust, with clear documentation and thoughtful use of recursion and memoization. The comprehensive test suite covers a wide range of scenarios, ensuring the correctness of the function. Overall, it's an excellent combination of efficiency, clarity, and thorough validation. I have a few suggestions for improvement, which I will add directly to your code.

File: solutions/distinct_subsequences.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            The source string in which to find subsequences.
+        t: str
+            The target string to match as a subsequence.
+
+    Returns:
+        int: The number of distinct subsequences of t in s.
+
+    Raises:
+        AssertionError: If the inputs are not strings.
+
```
Comment (Zeinab15) [inline solutions/distinct_subsequences.py:58]: It would be more appropriate to use TypeError instead of AssertionError in this case, as the inputs are being checked using assert isinstance(s, str) and assert isinstance(t, str). Typically, TypeError is raised when the type of an argument does not match the expected type, especially in functions that perform type validation. Therefore, raising TypeError would be a better approach here.

**Review by Zeinab15 — COMMENTED** [review]
Some suggestions:

File: solutions/tests/test_distinct_subsequences.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        with self.assertRaises(AssertionError):
+            count_distinct_subsequences(s, t)
+
+    def test_both_non_strings(self):
+        """Test when both s and t are not strings."""
+        s = 123
+        t = 456
+        with self.assertRaises(AssertionError):
+            count_distinct_subsequences(s, t)
+
```
Comment (Zeinab15) [inline solutions/tests/test_distinct_subsequences.py:105]: Your tests are excellent and cover all key test cases, but additional tests could further enhance the coverage, such as strings with special characters, spaces between characters, and case sensitivity. For example, s = 'a b c', t = 'abc' for spaces, s = 'Abc', t = 'abc' for case sensitivity, and strings like '(@$$$&)' and '@$&)' for special characters.

**Review by Khusro-S — APPROVED** [review]
It took me a bit of time to fully understand how everything was implemented, but it's cleverly done 🙌

**Review by Zeinab15 — DISMISSED** [review]
Well done, approved!

**Review by GaiSamuel — APPROVED** [review]
Looks good!

**Review by GaiSamuel — APPROVED** [review]

**Review by Zeinab15 — APPROVED** [review]
